File: chapt5/chapt5_ws/src/demo_python_tf/demo_python_tf/dynamic_tf_broadcaster.py
import math

# Workaround: patch deprecated np.float for older transforms3d compatibility
import numpy as np

if not hasattr(np, "float"):
    np.float = np.float64

from cv2 import transform
import rclpy
from rclpy.node import Node
from tf2_ros import TransformBroadcaster
from geometry_msgs.msg import TransformStamped
from tf_transformations import quaternion_from_euler


class TFBroadcaster(Node):
    def __init__(self):
        super().__init__("tf_broadcaster")
        self.broadcaster_ = TransformBroadcaster(self)
        self.timer_ = self.create_timer(0.01, self.publish_tf)

    def publish_tf(self):
        transform = TransformStamped()
        transform.header.stamp = self.get_clock().now().to_msg()
        transform.header.frame_id = "camera_link"
        transform.child_frame_id = "bottle_link"
        transform.transform.translation.x = 0.2
        transform.transform.translation.y = 0.3
        transform.transform.translation.z = 0.5

        # Convert Euler angles to quaternion
        rotation_quat = quaternion_from_euler(0, 0, 0)
        transform.transform.rotation.x = rotation_quat[0]
        transform.transform.rotation.y = rotation_quat[1]
        transform.transform.rotation.z = rotation_quat[2]
        transform.transform.rotation.w = rotation_quat[3]

        # 发布静态坐标变换
        self.broadcaster_.sendTransform(transform)
        self.get_logger().info(f"TF:{transform}")


def main():
    rclpy.init()
    static_tf_broadcaster = TFBroadcaster()
    rclpy.spin(static_tf_broadcaster)
    rclpy.shutdown()

File: chapt5/chapt5_ws/src/demo_cpp_tf/src/dynamic_tf_broadcaster.cpp
#include <memory>
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "tf2_ros/transform_broadcaster.h" // 提供坐标广播器类
#include <chrono>

using namespace std::chrono_literals;

class DynamicTFBroadcaster : public rclcpp::Node
{
private:
      std::shared_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
      rclcpp::TimerBase::SharedPtr timer_;

public:
      DynamicTFBroadcaster() : Node("dynamic_tf_broadcaster")
      {
            tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
            timer_ = create_wall_timer(10ms, std::bind(&DynamicTFBroadcaster::publishTransform, this));
      }
      void publishTransform()
      {
            geometry_msgs::msg::TransformStamped transform;
            transform.header.stamp = this->get_clock()->now();
            transform.header.frame_id = "map";
            transform.child_frame_id = "base_link";
            transform.transform.translation.x = 2.0;
            transform.transform.translation.y = 2.0;
            transform.transform.translation.z = 2.0;
            tf2::Quaternion quat;
            quat.setRPY(0, 0, 30 * M_PI / 180);
            transform.transform.rotation = tf2::toMsg(quat);
            tf_broadcaster_->sendTransform(transform);
      }
};

int main(int argc, char **argv)
{
      rclcpp::init(argc, argv);
      auto node = std::make_shared<DynamicTFBroadcaster>();
      rclcpp::spin(node);
      rclcpp::shutdown();
      return 0;
}